common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -1458,87 +1460,78 @@ public SSLEngineResult wrap(ByteBuffer[] srcs, int srcsOffset, int srcsLength, B
 
             int bytesProduced = 0;
             int bytesConsumed = 0;
-        loop:
-            for (int i = srcsOffset; i < endOffset; ++i) {
-                final ByteBuffer src = srcs[i];
-                checkArgument(src != null, "srcs[%d] is null", i);
-                while (src.hasRemaining()) {
-                    final SSLEngineResult pendingNetResult;
-                    // Write plaintext application data to the SSL engine
-                    int result = writePlaintextData(
-                        src, min(src.remaining(), SSL3_RT_MAX_PLAIN_LENGTH - bytesConsumed));
-                    if (result > 0) {
-                        bytesConsumed += result;
-
-                        pendingNetResult = readPendingBytesFromBIO(
+            ByteBuffer outputBuffer = fillOutputBuffer(srcs, srcsOffset, endOffset);
+            if (outputBuffer.hasRemaining()) {
+                final SSLEngineResult pendingNetResult;
+                // Write plaintext application data to the SSL engine
+                int result = writePlaintextData(outputBuffer, outputBuffer.remaining());
```

Comment:
Actually, there's a nasty submit-blocking issue here.

IFF `result` is negative (or the method throws) AND we copied data from the original buffers into `outputBuffer` AND the error is non-fatal then we will drop application data (because we already modified the positions of the input buffers).

I don't have a good solution yet... If no data has been consumed then we could cache the original buffer state and reset them.  Otherwise it might be impossible and we'd have to kill the connection.

Reply:
Solved with the second commit, which doesn't actually mark the data in source buffers as consumed unless `writePlaintextData()` returns a positive result.

CL still not really ready for submission though as this has revealed we don't have any test coverage around the error cases in `wrap()`

Reply:
Tidied up the main test.  Still needs more unit tests for the buffer combining logic though.